nixos/networkd: Add routes from interfaces to [Route] section of .network file

Luflosi opened the pull request:
###### Motivation for this change
For motivation see #93635.
The difference to that PR is that this one resolves the merge conflict, adds more `route.options` and changes the existing test to work correctly with the new code.
Previously, the test would test the scripted network backend, even when executing `nix-build -A nixosTests.networking.networkd.routes`.
I added many of the other `route.options` as best as I could but I didn't test any of them because they exceed my networking knowledge. It's possible that I missed some options that would translate well from iproute to networkd. Or maybe I made a a mistake and I mapped some option to a slightly different networkd option.

I found a small difference between the iproute and networkd:

From `man ip route`, section `scope`:
If this parameter is omitted, ip assumes scope global for all gatewayed unicast routes, scope link for direct unicast and broadcast routes and scope host for local routes.

From `man systemd.network`, section `Scope=`:
For IPv4 route, defaults to "host" if Type= is "local" or "nat", and "link" if Type= is "broadcast", "multicast", or "anycast". In other cases, defaults to "global".

The default type is `unicast` for both. As you can see, in this case iproute sets the scope to `link` while networkd sets it to `global`. To make the test pass, I simply manually specify the scope in the test.


###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Luflosi committed:
nixos/networkd: Add routes from interfaces to [Route] section of .network file

Closes https://github.com/NixOS/nixpkgs/pull/93635.

rnhmjoj commented:
> Previously, the test would test the scripted network backend, even when executing nix-build -A nixosTests.networking.networkd.routes.

Thank you! It seems I had forgotten about networkd when implementing this.

I checked all the options for typos and mismatch and they look ok, however I noticed something:
`Metric`, `PreferredSource`, `Protocol`, `Scope`, `Source` and `Table` are technically not route options, in the sense used in the manual:
  > Other route options. See the symbol OPTIONS in the ip-route(8) manual page for the details.
I think they still work due to the syntax of `ip route add` and the way we build the command, but I haven't checked yet. In any case, the description needs to be reworded.

> The default type is unicast for both. As you can see, in this case iproute sets the scope to link while networkd sets it to global.

Uhm, I'm not sure how to deal with this. Should we pick and enforce the same default or leave it ambiguous? Either way, it should be documented.

samueldr commented:
Sorry @rnhmjoj, if it's about networking you can count me out :). I don't have the required expertise and knowledge.

rnhmjoj commented:
> Sorry @rnhmjoj, if it's about networking you can count me out :)

Oh, sorry I meant to ping @andir.

Luflosi commented:
> In any case, the description needs to be reworded

I can't think of a good description. Do you have a suggestion?

> Uhm, I'm not sure how to deal with this. Should we pick and enforce the same default or leave it ambiguous? Either way, it should be documented.

I think we should leave it ambiguous for now. Where would be a good place to document this?

rnhmjoj commented:
> I think we should leave it ambiguous for now. Where would be a good place to document this?

There's nothing in the manual about routes, so I think the description of `networking.interfaces.<name>.routes` is probably the only place.

Luflosi commented:
What do you think of
If the route type is the default "unicast", then `ip route` (and therefore the default scripted network backend) and networkd set the route scope differently. `ip route` sets it to `link` while networkd sets it to `global`. If you want consistency between the two implementations, set the route scope manually with `networking.interfaces.eth0.ipv4.routes = [{ options.scope = "link"; }]` for example.

rnhmjoj commented:
Sorry for the late reply, I have kind of forgotten about this. Yes, the explanation looks fine to me.

rnhmjoj commented:
ping

Luflosi commented:
Yeah sorry, might be another week before I get to it.

Luflosi commented:
Sorry for taking so long.
I added the documentation to the `networking.interfaces.<name>.ipv4.routes` option. Adding it to the ipv6 routes docs is not needed since "The value [Scope=] is not used for IPv6." according to `man systemd.network`.
I altered my previous proposed text slightly ("route scope" -> "scope of the route" and changed the example to `global`). Is the use of `<literal>` appropriate for achieving the same purpose as the ` in GitHub?
Do you think this is ready for merging now?

andir reviewed (changes requested):
Looks good on first look. Could you also add a changelog entry that documents the change in behaviour?

Luflosi commented:
I added a changelog entry under the "Backward Incompatibilities" section. Please read it and decide if it is good or if you can think of any improvements.

rnhmjoj commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -207,6 +207,15 @@ let
         type = with types; listOf (submodule (routeOpts 4));
         description = ''
           List of extra IPv4 static routes that will be assigned to the interface.
+          If the route type is the default <literal>unicast</literal>, then
```
I would mention `networking.useNetworkd` explicitely, like:
```
If the route type is the default <literal>unicast</literal>, then the scope is
set differently depending on the value of <option>networking.useNetworkd</option>:
the script-based backend sets it to <literal>link</literal>, while networkd sets
it to <literal>global</literal>.
```
Also, it may be a good idea to put the paragraph in a `<warning></warning>` block.

Luflosi replied:
Yes, your version is better.
I added the `<warning></warning>` block.
I kept the paragraph "If you want consistency between the two implementations [...]". Do you think I should remove that too?

rnhmjoj replied:
No, I think it's fine this way. Thank you.

rnhmjoj commented:
There's still this (minor) issue:
> Metric, PreferredSource, Protocol, Scope, Source and Table are technically not route options, in the sense used in the manual

Otherwise the change looks good.

Luflosi commented:
Yes but since there is no type checking being done on `networking.interfaces.<name>.ipv4.routes.*.options` and because of the way `ip route add` is called, I think it's entirely possible that users use this option to specify things like `Metric` in there. Should I add new options like `networking.interfaces.<name>.ipv4.routes.*.metric` instead?

rnhmjoj commented:
No, I think it's enough to change the description: mention that you can also specify metric, source and such.

Luflosi commented:
Like this?

rnhmjoj commented:
Yep, thank you! I'll take another look and hopefully merge it soon.

Luflosi commented:
Thank you. Do you know why the "Build NixOS manual" test is failing?

flokli commented:
You need to run some docbook generation pipeline, and commit both the
updated markdown and the corresponding docbook text.

rnhmjoj commented:
You must open a paragraph inside the <warning> block.
Don't ask me why, thankfully docbook well be gone soon.

diff --git a/nixos/modules/tasks/network-interfaces.nix b/nixos/modules/tasks/network-interfaces.nix
index e5457546907..ac51f130ba8 100644
--- a/nixos/modules/tasks/network-interfaces.nix
+++ b/nixos/modules/tasks/network-interfaces.nix
@@ -213,10 +213,10 @@ let
         type = with types; listOf (submodule (routeOpts 4));
         description = ''
           List of extra IPv4 static routes that will be assigned to the interface.
-          <warning>If the route type is the default <literal>unicast</literal>, then the scope is
+          <warning><para>If the route type is the default <literal>unicast</literal>, then the scope is
           set differently depending on the value of <option>networking.useNetworkd</option>:
           the script-based backend sets it to <literal>link</literal>, while networkd sets
-          it to <literal>global</literal>.</warning>
+          it to <literal>global</literal>.</para></warning>
           If you want consistency between the two implementations,
           set the scope of the route manually with
           <literal>networking.interfaces.eth0.ipv4.routes = [{ options.scope = "global"; }]</literal>


On 13-01-22, Luflosi wrote:
> Thank you. Do you know why the "Build NixOS manual" test is failing?
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/NixOS/nixpkgs/pull/144590#issuecomment-1012129681
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

Luflosi commented:
Yeah, that was it, thanks @rnhmjoj.

rnhmjoj reviewed (approved):
I double checked the changes, built the manual and run the networking (tested/networkd) tests. It seems everything is in order.

rnhmjoj commented:
@andir @flokli can you finish your review?

rnhmjoj commented:
@Luflosi can you fix the conflict?

Luflosi commented:
Done. I followed the advice in that file: "To avoid merge conflicts, consider adding your item at an arbitrary place in the list instead.".

rnhmjoj commented:
@GrahamcOfBorg test networking.networkd.routes networking.networkd.routes

ncfavier commented:
Broke the networking.networkd.static test

Luflosi commented:
```patch
--- a/nixos/modules/tasks/network-interfaces-systemd.nix
+++ b/nixos/modules/tasks/network-interfaces-systemd.nix
@@ -92,7 +92,7 @@ in
             };
           };
         });
-        networks."40-${i.name}" = mkMerge [ (genericNetwork mkDefault) {
+        networks."40-${i.name}" = mkMerge [ (genericNetwork mkMerge) {
           name = mkDefault i.name;
           DHCP = mkForce (dhcpStr
             (if i.useDHCP != null then i.useDHCP else false));
```
seems to fix it. Is that a good fix or does it break something else?

ncfavier commented:
Can you open a PR so we can run the tests?

Luflosi commented:
> Applying your patch as-is gives

Sorry, my bad.

I think the problem is, that setting `defaultGateway` does this
```nix
systemd.network.networks."40-${i.name}".routes = mkDefault [
  {
    routeConfig = {
      Gateway = gateway;
      GatewayOnLink = false;
    };
  }
];
```
while the code I added (if `networking.interfaces.<name>.ipv4.routes` is not set) does this:
```nix
systemd.network.networks."40-${i.name}".routes = [ ];
```
(If I understand the code correctly).
This means that the latter overrides the former. Can we just remove the `mkDefault` above, so the two lists are merged?
Like this:
```patch
--- a/nixos/modules/tasks/network-interfaces-systemd.nix
+++ b/nixos/modules/tasks/network-interfaces-systemd.nix
@@ -66,7 +66,7 @@ in
           let gateway = optional (cfg.defaultGateway != null && (cfg.defaultGateway.address or "") != "") cfg.defaultGateway.address
             ++ optional (cfg.defaultGateway6 != null && (cfg.defaultGateway6.address or "") != "") cfg.defaultGateway6.address;
           in optionalAttrs (gateway != [ ]) {
-            routes = override [
+            routes = [
               {
                 routeConfig = {
                   Gateway = gateway;
```

ncfavier commented:
Ok, that sounds right to me and passes the tests. You can implement it as `genericNetwork id` (is that argument really useful btw?)

Luflosi commented:
I'll create a PR with your suggestion.

> is that argument really useful btw?

Not entirely sure, I honestly don't understand the code well enough. It was presumably added for a reason.

Luflosi commented:
PR: #156083.
